### ISSUE TITLE
Additional External Editors to "Open With" Menu

### ISSUE THREAD

**Issue by rat-moonshine** [Moonshine-IDE/Moonshine-IDE#993]
From @JustinProminic ,

-----------------------

I have spent the past several hours researching a variety of software packages for Mac to determine additional ones that I believe need to be tied into Moonshine as additional external editors.

We are not going to be adding these to Moonshine SDK Installer.

Please perform the following steps:

1) Download each of these tools so you have them on your machine if you do not already:

- [x] IntelliJ - https://www.jetbrains.com/
- [x] MyEclipse - https://www.genuitec.com/products/myeclipse/download/
- [x] NetBeans for Mac - https://netbeans.apache.org/download/nb126/nb126.html
- [x] BlueJ - https://www.bluej.org/ - Note that BlueJ does not have an installer, so please just assume in the external editor default location that the user places it in /Applications/BlueJ
- [x] JGrasp - https://www.jgrasp.org/ - There is some chance I may later ask for JGrasp to be part of the Moonshine SDK Installer but for now we will just put it as an external editor.
- [x] Brackets.io - https://brackets.io/
- [x] MultiMarkdown Composer - https://apps.apple.com/us/app/multimarkdown-composer-4/id1275176220?ls=1&mt=12
or https://multimarkdown.com/composer4/
- [x] Balsamiq Wireframes for Desktop which should be enabled if the filename selected ends in ".bmpr" https://balsamiq.com/wireframes/desktop/
- [x] JarZilla which should be enabled if the filename selected ends in ".jar": https://github.com/depsypher/jarzilla/releases/tag/v0.1.6 (https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1047)
- [x] Java Decompiler - https://java-decompiler.github.io

2) Add entries (in alphabetical order along with the existing ones) to the External Editors section of Moonshine:

3) Add them to the right click Open With menu.   To do this properly so that they open the selected file, you will need to experiment with the command line parameters for passing a file.  I assume, but am not totally sure, that each of the above editors/IDEs do actually support passing a single file as a command line to open them.

Once this task is done, the above should appear here:

![image](https://user-images.githubusercontent.com/24370870/158114852-5babec2b-3e00-431b-8cb0-f59adbfd24f9.png)

![image](https://user-images.githubusercontent.com/24370870/158114783-f0a859a3-2cbf-4b12-a58f-8ee0cf6a1651.png)

-----------------------------------------

Additionally, based on the given requirement, it seems we'll need filtration capability based on file type(s) to enable/disable editors. Currently, there is no such filtration but we'll need one.

**Comment by rat-moonshine** [comment]
Covers/Added:

- IntelliJ IDEA
- MyEclipse
- Balsamiq Mockups

As an additional to the original requirement, it seems needs an additional filtration, as well. The new filtration would also responsible to editor's enable/disable based on the targeted file type, i.e. Balsamiq Mockups should enable on `.bmpr` only. 

The above filtration is now in place.

**Comment by rat-moonshine** [comment]
This issue is ready for a test. 

**Note:** Some editors may not auto-detect even they're installed. This true only for the editors which installs with version values label in their installed directory or to the executable. Moonshine do not has any complicated AI to determine the directory/executable label but maintains a simple [XML](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/42db167136c50ad22ca6be55801a8dccd7fc4481/ide/MoonshineSharedCore/src/elements/data/DefaultExternalEditors.xml) to valid/possible paths. User may need to adjust such editors' paths manually in Moonshine. 

Please, check.

**Comment by rat-moonshine** [comment]
Editor | Auto-detect (Win) | Auto-detect (OSX) | Type-restriction | Opens File (Win) | Opens File (OSX)
-------|----------|------------------------|-----------------|-----------------|-------------
IntelliJ IDEA | No (version labelled directory) | Yes | None | ? | Yes
MyEclipse | Yes | Yes | None | ? | Yes
Balsamiq Mockups | Yes | Yes | `bmpr` | ? | Yes (works for same version created files)
BlueJ | Yes | Yes | `bluej` | ? | Yes
jGRASP | Yes | Yes | None | ? | Yes
Jarzilla | NA | Yes | `jar` | NA | ? (not opens on Catalina)
Java Decompiler | Yes | Yes | `jar`, `class` | Yes | ? (not opens on Catalina)
Brackets | Yes | Yes | None | ? | Yes
NetBeans | Yes | Yes | None | ? | ?
Multimarkdown Composer | NA | Yes | None | NA | Yes

**Comment by Aszusz** [comment]
Both Jarzilla and Java Decompiler don't open for me because of Java problems. Jarzilla requires Java 1.7+ and Java Decompiler requires Java 1.8+. I installed openjdk8 via mac ports and set up my JAVA_HOME:
```
aszusz@adrians-mac-mini ~ % java -version 
openjdk version "1.8.0_312"
OpenJDK Runtime Environment (Temurin)(build 1.8.0_312-b07)
OpenJDK 64-Bit Server VM (Temurin)(build 25.312-b07, mixed mode)
aszusz@adrians-mac-mini ~ % echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/openjdk8-temurin/Contents/Home
```
and they still won't open. Jarzilla seems to be severely outdated and I'd suggest dropping support for it. Java Decompiler may still be worth the hassle.

NetBeans works for me on macOS. When I right-click a java class and try to open it with NetBeans the application starts but without the file opened. So we need to tweak passing the file as a parameter and it should be fine.

**Comment by JoelProminic** [comment]
Jarzilla and Java Decompiler work for me from the command line:

    % open -a /Applications/Jarzilla.app ~/Documents/Flex/temp_workspace/TestJavaMaven/target/TestJavaMaven-0.1.0-SNAPSHOT.jar
    % less /Applications/JD-GUI.app/Contents/Info.plist 

    % open /Applications/JD-GUI.app 
    % open -a /Applications/JD-GUI.app ~/Documents/Flex/temp_workspace/TestJavaMaven/target/TestJavaMaven-0.1.0-SNAPSHOT.jar


However, Jarzilla and Java Decompiler failed from Moonshine with this error (seen in with the "Report" button):

    Dyld Error Message:
      Library not loaded: @rpath/libjli.dylib
      Referenced from: /Users/USER/Downloads/*/java
      Reason: no suitable image found.  Did find:
            file system sandbox blocked open() of '/Users/joelanderson/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib'
            /Users/joelanderson/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib: stat() failed with errno=1


This looks like a sandbox error.  I also considered that it could be a error with DYLD_LIBRARY_PATH not being preserved, but I see that the `.dylib` files were detected, but could not be opened.


I installed NetBeans from the website linked in the description, and I got version 12.6 (/Applications/Apache NetBeans 12.6.app).  Moonshine was expecting version 13.


I found that Apache NetBeans was not automatically enabled after I set the path.

Once I fixed these items, I confirmed that I could open a .java file in NetBeans on macOS Catalina.  It worked regardless of whether NetBeans was already open.

**Comment by JoelProminic** [comment]
@rat-moonshine reported that jGrasp was working earlier, so we should compare that application to Jarzilla and Java Decompiler.

@JustinProminic would also like to try copying Jarzilla and Java Decompiler to try to resolve the sandbox error.  We shouldn't distribute it this way, but it would be an interesting test for debugging.

We would like to get all of the editors working before we close this issue.

**Comment by rat-moonshine** [comment]
I give local tests to the idea to include Jarzilla and Java Decompiler as Moonshine resource, as @JustinProminic suggested, but that returned me same error as @JoelProminic [reported](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/993#issuecomment-1091991090).

To ensure the application uses its run-time Java dependencies correctly I test the application as an installed package. Following were my test procedures:

- Include `Jarzilla.app` and `JD-GUI.app` as part of Moonshine resource file - I see I probably can not just copy the executable `.jar` file because the `.app` file also consists with many other files possibly are support files to the executable `.jar` under `$app/Contents/Resources`, I chose to copy the complete `.app` files
- I remove any possible extended-attributes from the `.app` files
- Modified editor trigger codes to use from bundled location for Jarzilla and Java Decompiler
- Packaged the application using Ant build script and test

I see same error window while they triggers from Moonshine-IDE:
![image](https://user-images.githubusercontent.com/24370870/165688813-38fed82d-c49c-40ee-93cb-b0057fc10ff4.png)

along with same error-message under "Report" tab:
```
Dyld Error Message:
  Library not loaded: @rpath/libjli.dylib
  Referenced from: /Users/USER/Downloads/*/java
  Reason: no suitable image found.  Did find:
	file system sandbox blocked open() of '/Users/devsena/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib'
	/Users/devsena/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib: stat() failed with errno=1

```

These two applications continue not to open in my standalone tests on Catalina.

**Comment by JoelProminic** [comment]
@JustinProminic reported that the BlueJ download URL pointed to the Balsamiq Mockups page instead.  It should point to:  https://www.bluej.org

Note that the download page provides a zip rather than a .pkg or .app.   The structure is like this:

    BlueJ 5.0.3/
        Bluej.app
        examples/

If we assume the user copies the entire directory to /Applications, then default path would be `/Applications/BlueJ 5.0.3/BlueJ.app`, which will become obsolete after each update.  We could also choose `/Applications/BlueJ/BlueJ.app`, but this would be wrong unless the user decided to rename the directory.

I would prefer to assume that the user copied the application only, and keep the current default path as `/Applications/BlueJ.app`.

**Comment by JoelProminic** [comment]
I did a little more investigation for JarZilla.  the error looks like this:

    Process:               java [63139]
    Path:                  /Applications/Jarzilla.app/Contents/MacOS/universalJavaApplicationStub
    ...

    Dyld Error Message:
      Library not loaded: @rpath/libjli.dylib
      Referenced from: /Users/USER/Downloads/*/java
      Reason: no suitable image found.  Did find:
            file system sandbox blocked open() of '/Users/-user-/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib'
            /Users/-user-/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/../lib/jli/libjli.dylib: stat() failed with errno=1

I confirmed that libjli.dylib exists at the indicated path.

Looking into the "file system sandbox blocked open()" error, I [found](https://stackoverflow.com/questions/44627957/mac-os-file-system-sandbox-blocked-open) that this is related to the Files and Folder/Full Disk Access options in macOS security.  I tried adding "Full Disk Access" entries for the following, but this didn't help:
- Jarzilla.app
- `/Applications/Jarzilla.app/Contents/MacOS/universalJavaApplicationStub`
- `/Users/-user-/Downloads/MoonshineSDKs/Java/openjdk-11.0.10/bin/`

Another suggestion in the above page was to copy the libraries to `/Applications/Jarzilla.app/Contents/Frameworks/`.  We could try this, but don't think it is a practical solution for us.

Testing from Terminal, I noticed that Jarzilla was instead using the system JDKs at `/Library/Java/JavaVirtualMachines`.  This is probably retrieved with a similar method to the `/usr/libexec/java_home` command (note that this also meant that the application was broken from Terminal and Finder when I had a newer SDK installed (temurin-17.jdk`)).    I see this even if I have JAVA_HOME set in the environment.

I tried setting the Moonshine Java SDK to a `/Library/Java/JavaVirtualMachines` path.  With JDK 8, this worked properly! (but it did not work for JDK 11).  We could try overriding JAVA_HOME in this case, but this requires that an appropriate JDK is installed.  We can get this path with:

    % /usr/libexec/java_home -v 1.8
    /Library/Java/JavaVirtualMachines/jdk1.8.0_201.jdk/Contents/Home

**Comment by JoelProminic** [comment]
From @JustinProminic,

>We might be able to wrapper starting the JAR as follows:
>
>https://apple.stackexchange.com/questions/191609/how-to-create-a-app-folder-from-an-executable-jar
>
>
>--do shell script "java -jar /Contents/execute.jar"
>
>
>and all you would need to do that is allow .JAR files to be selected in the External Editors.  Currently only .App files can be selected.
>
>
>
>Other options are adding instructions such as:
>
>https://github.com/Jorl17/jar2app
>
>from:
>
>https://stackoverflow.com/questions/2695214/convert-jar-to-an-osx-executable

I did a test, and found that this worked for Java Decompiler, but not JarZilla:

    % $JAVA_HOME/bin/java -jar /Applications/JD-GUI.app/Contents/Resources/Java/jd-gui-1.6.6-min.jar <project>/build/libs/JavaAgentIMOSValidator.jar
    # Application opens and works!
    % $JAVA_HOME/bin/java -jar /Applications/Jarzilla.app/Contents/Resources/Java/jarzilla.jar 
    Exception in thread "main" java.lang.NoClassDefFoundError: ch/randelshofer/quaqua/QuaquaManager
            at com.googlecode.jarzilla.Jarzilla.<init>(Jarzilla.java:83)
            at com.googlecode.jarzilla.Jarzilla.main(Jarzilla.java:485)
    Caused by: java.lang.ClassNotFoundException: ch.randelshofer.quaqua.QuaquaManager
            at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
            at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
            at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
            at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
            ... 2 more


I tried to workaround this by specifying the classpath and the main class instead.  To get the main class, you need to unzip the main JAR (or use JD-GUI), and look for this line in `META-INF/MANIFEST.MF`

    Main-Class: com.googlecode.jarzilla.Jarzilla


Then you can run the application like this.  The classpath is a ':' separated list of JAR filesin the same directory:

    cd /Applications/Jarzilla.app/Contents/Resources/Java
    # java -cp '<classpath>' <mainclass> <target_jar>
    $JAVA_HOME/bin/java -cp 'forms-1.2.1.jar:jarzilla.jar:jsyntaxpane-0.9.4.jar:mac_widgets.jar:quaqua-filechooser-only.jar:schlepit.jar:swingx-1.6.jar' com.googlecode.jarzilla.Jarzilla <target_jar>

However, I found that this only opened the application.  The specified `target_jar` was not opened.o

If we want to proceed with this, we'll need to spend some time making a generic script to handle this logic.  Even then, I would not expect this to work for every Java project (since dependencies could be in unexpected locations).

**Comment by JoelProminic** [comment]
From @JustinProminic

>We need to cleanup the install instructions for Java Decompiler.
>
>It is unclear how to get it to this point the Moonshine IDE calls for in default of /Applications/JD-GUI.app :
>
>![image](https://user-images.githubusercontent.com/24532731/171293773-8f799198-6c95-4d2d-b2cd-85e066635148.png)
>
>Specifically, the download page has a Mac installer but doesn't install to the above folder.
>
>When I downloaded the JAR release, it will run and creates a config file.

Checking this, I found that the Java option for [JD-GUI](https://java-decompiler.github.io/) was a TAR file (jd-gui-osx-1.6.6.tar).  I extracted this file, and it contained JD-GUI.app (and other files).  I copied JD-GUI.app to /Applications.

If we want to use the executable JAR solution above, we could recommend the user downloads the JAR instead.

How much guidance do we want to give for these installs?  My understanding was that we letting the user figure out the installation for the external editors, and treating this as an advanced feature.

**Comment by rat-moonshine** [comment]
I have created a separate issue to track Jarzilla problem - https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1047.

**Comment by JoelProminic** [comment]
We have separate issues for the remaining Java problems.  Otherwise this issue looks good.
